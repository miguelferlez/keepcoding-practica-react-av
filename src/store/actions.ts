import { AxiosError } from "axios";
import type { AppThunk } from "./index";
import type { Advert } from "@/pages/adverts/types";
import type { Credentials } from "@/pages/auth/types";

//#region Auth

type AuthLoginFulfilled = {
  type: "auth/login/fulfilled";
};
type AuthLoginPending = {
  type: "auth/login/pending";
};
type AuthLoginRejected = {
  type: "auth/login/rejected";
  payload: AxiosError<{ message: string }>;
};
type AuthLogout = {
  type: "auth/logout";
};

const authLoginFulfilled = (): AuthLoginFulfilled => ({
  type: "auth/login/fulfilled",
});
const authLoginPending = (): AuthLoginPending => ({
  type: "auth/login/pending",
});
const authLoginRejected = (
  error: AxiosError<{ message: string }>,
): AuthLoginRejected => ({
  type: "auth/login/rejected",
  payload: error,
});

export function authLogin(
  credentials: Credentials,
  remember: boolean,
): AppThunk<Promise<void>> {
  return async function (dispatch, _getState, { api, router }) {
    dispatch(authLoginPending());

    try {
      await api.auth.login(credentials, remember);
      dispatch(authLoginFulfilled());
      router.navigate(router.state.location.state?.from ?? "/", {
        replace: true,
      });
    } catch (error) {
      if (error instanceof AxiosError) {
        dispatch(authLoginRejected(error));
      }
      throw error;
    }
  };
}

export function authLogout(): AuthLogout {
  return { type: "auth/logout" };
}

//#endregion

//#region Ui

type UiResetError = {
  type: "ui/reset-error";
};

export const uiResetError = (): UiResetError => {
  return { type: "ui/reset-error" };
};

//#endregion

//#region Adverts

type AdvertsLoadedFulfilled = {
  type: "adverts/loaded/fulfilled";
  payload: Advert[];
};
type AdvertsLoadedPending = {
  type: "adverts/loaded/pending";
};
type AdvertsLoadedRejected = {
  type: "adverts/loaded/rejected";
  payload: AxiosError<{ message: string }>;
};

const advertsLoadedFulfilled = (adverts: Advert[]): AdvertsLoadedFulfilled => ({
  type: "adverts/loaded/fulfilled",
  payload: adverts,
});
const advertsLoadedPending = (): AdvertsLoadedPending => ({
  type: "adverts/loaded/pending",
});
const advertsLoadedRejected = (
  error: AxiosError<{ message: string }>,
): AdvertsLoadedRejected => ({
  type: "adverts/loaded/rejected",
  payload: error,
});

export function advertsLoaded(): AppThunk<Promise<void>> {
  return async function (dispatch, getState, { api }) {
    const state = getState();

    if (state.adverts.loaded) {
      return;
    }

    try {
      dispatch(advertsLoadedPending());
      const adverts = await api.adverts.getAdverts();
      setTimeout(() => {
        dispatch(advertsLoadedFulfilled(adverts));
      }, 1000);
    } catch (error) {
      if (error instanceof AxiosError) {
        dispatch(advertsLoadedRejected(error));
      }
      throw error;
    }
  };
}

//#endregion

//#region Adverts tags

type AdvertsTagsFulfilled = {
  type: "adverts/tags/fulfilled";
  payload: string[];
};

type AdvertsTagsPending = {
  type: "adverts/tags/pending";
};

type AdvertsTagsRejected = {
  type: "adverts/tags/rejected";
  payload: AxiosError<{ message: string }>;
};

const advertsTagsFulfilled = (tags: string[]): AdvertsTagsFulfilled => ({
  type: "adverts/tags/fulfilled",
  payload: tags,
});

const advertsTagsPending = (): AdvertsTagsPending => ({
  type: "adverts/tags/pending",
});

const advertsTagsRejected = (
  error: AxiosError<{ message: string }>,
): AdvertsTagsRejected => ({
  type: "adverts/tags/rejected",
  payload: error,
});

export function advertsTags(): AppThunk<Promise<void>> {
  return async function (dispatch, _getState, { api }) {
    dispatch(advertsTagsPending());

    try {
      const tags = await api.adverts.getAdvertsTags();
      dispatch(advertsTagsFulfilled(tags));
    } catch (error) {
      if (error instanceof AxiosError) {
        dispatch(advertsTagsRejected(error));
      }
      throw error;
    }
  };
}

//#endregion

//#region Adverts detail

type AdvertsDetailFulfilled = {
  type: "adverts/detail/fulfilled";
  payload: Advert;
};
type AdvertsDetailPending = {
  type: "adverts/detail/pending";
};
type AdvertsDetailRejected = {
  type: "adverts/detail/rejected";
  payload: AxiosError<{ message: string }>;
};

const advertsDetailFulfilled = (advert: Advert): AdvertsDetailFulfilled => ({
  type: "adverts/detail/fulfilled",
  payload: advert,
});
const advertsDetailPending = (): AdvertsDetailPending => ({
  type: "adverts/detail/pending",
});
const advertsDetailRejected = (
  error: AxiosError<{ message: string }>,
): AdvertsDetailRejected => ({
  type: "adverts/detail/rejected",
  payload: error,
});

export function advertsDetail(advertId: string): AppThunk<Promise<void>> {
  return async function (dispatch, _getState, { api }) {
    dispatch(advertsDetailPending());

    try {
      const advert = await api.adverts.getAdvertById(advertId);
      dispatch(advertsDetailFulfilled(advert));
    } catch (error) {
      if (error instanceof AxiosError) {
        dispatch(advertsDetailRejected(error));
      }
      throw error;
    }
  };
}

//#endregion

export type Actions =
  | AuthLoginFulfilled
  | AuthLoginPending
  | AuthLoginRejected
  | AuthLogout
  | UiResetError
  | AdvertsLoadedFulfilled
  | AdvertsLoadedPending
  | AdvertsLoadedRejected
  | AdvertsTagsFulfilled
  | AdvertsTagsPending
  | AdvertsTagsRejected
  | AdvertsDetailFulfilled
  | AdvertsDetailPending
  | AdvertsDetailRejected;
