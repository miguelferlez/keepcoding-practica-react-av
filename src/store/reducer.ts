import type { AxiosError } from "axios";
import type { Actions } from "./actions";
import type { Advert } from "../pages/adverts/types";

export type State = {
  auth: boolean;
  ui: {
    pending: boolean;
    error: AxiosError<{ message: string }> | null;
  };
  adverts: { loaded: boolean; data: Advert[] };
};

const defaultState: State = {
  auth: false,
  ui: {
    pending: false,
    error: null,
  },
  adverts: { loaded: false, data: [] },
};

export function auth(
  state = defaultState.auth,
  action: Actions,
): State["auth"] {
  switch (action.type) {
    case "auth/login/fulfilled":
      return true;
    case "auth/logout":
      return false;
    default:
      return state;
  }
}

export function ui(state = defaultState.ui, action: Actions): State["ui"] {
  switch (action.type) {
    case "auth/login/pending":
      return { pending: true, error: null };
    case "auth/login/fulfilled":
      return { pending: false, error: null };
    case "auth/login/rejected":
      return { pending: false, error: action.payload };
    case "adverts/loaded/pending":
      return { pending: true, error: null };
    case "adverts/loaded/fulfilled":
      return { pending: false, error: null };
    case "adverts/loaded/rejected":
      return { pending: false, error: action.payload };
    case "ui/reset-error":
      return { ...state, error: null };
    default:
      return state;
  }
}

export function adverts(
  state = defaultState.adverts,
  action: Actions,
): State["adverts"] {
  switch (action.type) {
    case "adverts/loaded/fulfilled":
      return { loaded: true, data: action.payload };
    default:
      return state;
  }
}
