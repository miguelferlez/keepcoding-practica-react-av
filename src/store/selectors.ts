import type { RootState } from ".";

export const getIsLogged = (state: RootState) => state.auth;

export const getUi = (state: RootState) => state.ui;
