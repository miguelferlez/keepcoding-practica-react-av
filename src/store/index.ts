import { combineReducers, createStore, applyMiddleware } from "redux";
import * as reducers from "./reducer";
import { useDispatch, useSelector } from "react-redux";
import { composeWithDevTools } from "@redux-devtools/extension/";
import * as thunk from "redux-thunk";
import type { Actions } from "./actions";
import type { createBrowserRouter } from "react-router";
import * as auth from "../services/auth";
import * as adverts from "../services/adverts";

const rootReducer = combineReducers(reducers);
type Router = ReturnType<typeof createBrowserRouter>;
type ExtraArgument = {
  api: { auth: typeof auth; adverts: typeof adverts };
  router: Router;
};

export default function configureStore(
  preloadedState: Partial<reducers.State>,
  router: Router,
) {
  const store = createStore(
    rootReducer,
    preloadedState as never,
    composeWithDevTools(
      applyMiddleware(
        thunk.withExtraArgument<reducers.State, Actions, ExtraArgument>({
          api: { auth, adverts },
          router,
        }),
      ),
    ),
  );

  return store;
}

export type AppStore = ReturnType<typeof configureStore>;
export type AppGetState = AppStore["getState"];
export type RootState = ReturnType<AppGetState>;
export type AppDispatch = AppStore["dispatch"];
export type AppThunk<ReturnType = void> = thunk.ThunkAction<
  ReturnType,
  RootState,
  ExtraArgument,
  Actions
>;

export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();
