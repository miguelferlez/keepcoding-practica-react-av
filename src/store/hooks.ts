import { useAppDispatch, useAppSelector } from ".";
import type { Credentials } from "../pages/auth/types";
import { authLogin, authLogout, uiResetError } from "./actions";
import { getIsLogged } from "./selectors";

export function useAuth() {
  return useAppSelector(getIsLogged);
}

export function useLoginAction() {
  const dispatch = useAppDispatch();
  return function (credentials: Credentials, remember: boolean) {
    return dispatch(authLogin(credentials, remember));
  };
}
export function useLogoutAction() {
  const dispatch = useAppDispatch();
  return function () {
    return dispatch(authLogout());
  };
}

export function useUiResetErrorAction() {
  const dispatch = useAppDispatch();
  return function () {
    return dispatch(uiResetError());
  };
}
