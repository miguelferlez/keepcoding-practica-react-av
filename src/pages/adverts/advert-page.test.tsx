import { AxiosError } from "axios";
import { Provider } from "react-redux";
import { MemoryRouter, Route, Routes } from "react-router";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import AdvertPage from "./advert-page";
import type { Advert } from "./types";
import type { RootState } from "@/store";
import { advertsDeleted, advertsDetail, uiResetError } from "@/store/actions";

vi.mock("@/store/actions");
vi.mock("date-fns", () => {
  return {
    formatDistanceToNow: () => "X days",
  };
});

describe("AdvertPage", () => {
  const state: RootState = {
    auth: false,
    ui: {
      pending: { auth: false, adverts: false, tags: false },
      error: null,
    },
    adverts: { loaded: false, data: [], tags: [] },
  };

  const advert = {
    id: "1",
    createdAt: "2025-08-10T10:25:58.000Z",
    name: "test",
    sale: false,
    price: 0,
    tags: ["work", "lifestyle"],
    photo: null,
  };

  const store = {
    getState: () => state,
    subscribe: () => {},
    dispatch: () => {},
  };

  const renderComponent = (
    advert: Advert,
    error?: AxiosError<{ message: string }>,
    pending: Partial<RootState["ui"]["pending"]> = {},
  ) => {
    state.adverts.data = [advert];
    state.ui.error = error || null;
    state.ui.pending = { ...state.ui.pending, ...pending };

    return render(
      // @ts-expect-error: store completed config not needed
      <Provider store={store}>
        <MemoryRouter initialEntries={[`/adverts/${advert.id}`]}>
          <Routes>
            <Route path="/adverts/:advertId" element={<AdvertPage />} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    );
  };

  afterEach(() => {
    vi.clearAllMocks();
  });

  it("should dispatch loader when pending", () => {
    const { container } = renderComponent(advert, undefined, { adverts: true });

    expect(container).toMatchSnapshot();
    expect(screen.getByRole("status")).toBeInTheDocument();
  });

  it("should dispatch advert details", () => {
    const { container } = renderComponent(advert, undefined, {
      adverts: false,
    });

    expect(container).toMatchSnapshot();
    expect(advertsDetail).toHaveBeenCalledWith(advert.id);
    expect(screen.getByText(advert.name)).toBeInTheDocument();
    expect(screen.getByText(`$${advert.price.toFixed(2)}`)).toBeInTheDocument();
    expect(screen.getByText("Buy")).toBeInTheDocument();
    expect(screen.getByText(`#${advert.tags[0]}`)).toBeInTheDocument();
    expect(screen.getAllByText(/#/)).toHaveLength(2);
  });

  it("should dispatch advert deletion when button and modal clicked", async () => {
    const { container } = renderComponent(advert, undefined, {
      adverts: false,
    });
    const deleteButton = screen.getByRole("button");

    expect(container).toMatchSnapshot();
    // open modal
    await userEvent.click(deleteButton);
    const modal = screen.getByRole("dialog");
    expect(modal).toBeInTheDocument();
    // confirm modal
    const confirmModalButton = screen.getByRole("button", { name: /yes/i });
    await userEvent.click(confirmModalButton);
    expect(advertsDeleted).toHaveBeenCalledWith(advert.id);
  });

  it("should cancel advert deletion when button clicked but modal cancel", async () => {
    const { container } = renderComponent(advert, undefined, {
      adverts: false,
    });
    const deleteButton = screen.getByRole("button");

    expect(container).toMatchSnapshot();
    // open modal
    await userEvent.click(deleteButton);
    const modal = screen.getByRole("dialog");
    expect(modal).toBeInTheDocument();
    // close modal
    const closeModalButton = screen.getByRole("button", { name: /close/ });
    await userEvent.click(closeModalButton);
    expect(advertsDeleted).not.toHaveBeenCalled();
    expect(modal).not.toBeInTheDocument();
  });

  it("should dispatch alert component with error message", async () => {
    const error = new AxiosError<{ message: string }>("Error message");
    const { container } = renderComponent(advert, error, { adverts: false });
    const alert = screen.getByRole("alert");

    expect(container).toMatchSnapshot();
    expect(alert).toHaveTextContent(error.message);
    await userEvent.click(alert);
    expect(uiResetError).toHaveBeenCalled();
  });
});
