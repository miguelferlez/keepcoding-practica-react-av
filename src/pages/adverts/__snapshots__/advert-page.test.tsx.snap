// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AdvertPage > should cancel advert deletion when button clicked but modal cancel 1`] = `
<div>
  <div
    class="wrapper only:mt-8"
  >
    <h2>
      Advert Details
    </h2>
    <div
      class="flex flex-col gap-4 md:flex-row"
    >
      <div
        class="md:flex-1"
      >
        <div
          class="mb-4 h-[370px] overflow-hidden rounded-lg bg-[#384152]"
        >
          <img
            alt="Product Image"
            class="h-full w-full object-cover"
            src="/src/assets/placeholder.webp"
          />
        </div>
        <div
          class="flex items-center justify-end"
        >
          <button
            class="btn btn-destructive"
          >
            <span
              class="-mb-0.75"
            >
              Delete this advert
            </span>
          </button>
        </div>
      </div>
      <div
        class="rounded-lg border border-neutral-500/15 p-4 md:flex-1"
      >
        <h3>
          test
        </h3>
        <div
          class="divide-b mb-2 flex gap-2 pb-2"
        >
          <span
            class="font-bold"
          >
            Price:
          </span>
          <span
            class="font-bold text-teal-600"
          >
            $
            0.00
          </span>
        </div>
        <div
          class="divide-b mb-2 flex gap-2 pb-2"
        >
          <span
            class="font-bold"
          >
            Wants to:
          </span>
          <span
            class=""
          >
            Buy
          </span>
        </div>
        <div
          class="divide-b mb-2 flex flex-wrap gap-x-2 gap-y-0.5 pb-2"
        >
          <span
            class="font-bold"
          >
            Tags:
          </span>
          <span>
            #
            work
          </span>
          <span>
            #
            lifestyle
          </span>
        </div>
        <div>
          <time
            datetime="2025-08-10T10:25:58.000Z"
          >
            <p
              class="text-primary/50"
            >
              This advert was uploaded
               
              7 days
               ago.
            </p>
          </time>
        </div>
      </div>
    </div>
    <div
      class="fixed top-11.5 left-1/2 z-50 -translate-x-1/2"
    />
  </div>
</div>
`;

exports[`AdvertPage > should dispatch advert deletion when button and modal clicked 1`] = `
<div>
  <div
    class="wrapper only:mt-8"
  >
    <h2>
      Advert Details
    </h2>
    <div
      class="flex flex-col gap-4 md:flex-row"
    >
      <div
        class="md:flex-1"
      >
        <div
          class="mb-4 h-[370px] overflow-hidden rounded-lg bg-[#384152]"
        >
          <img
            alt="Product Image"
            class="h-full w-full object-cover"
            src="/src/assets/placeholder.webp"
          />
        </div>
        <div
          class="flex items-center justify-end"
        >
          <button
            class="btn btn-destructive"
          >
            <span
              class="-mb-0.75"
            >
              Delete this advert
            </span>
          </button>
        </div>
      </div>
      <div
        class="rounded-lg border border-neutral-500/15 p-4 md:flex-1"
      >
        <h3>
          test
        </h3>
        <div
          class="divide-b mb-2 flex gap-2 pb-2"
        >
          <span
            class="font-bold"
          >
            Price:
          </span>
          <span
            class="font-bold text-teal-600"
          >
            $
            0.00
          </span>
        </div>
        <div
          class="divide-b mb-2 flex gap-2 pb-2"
        >
          <span
            class="font-bold"
          >
            Wants to:
          </span>
          <span
            class=""
          >
            Buy
          </span>
        </div>
        <div
          class="divide-b mb-2 flex flex-wrap gap-x-2 gap-y-0.5 pb-2"
        >
          <span
            class="font-bold"
          >
            Tags:
          </span>
          <span>
            #
            work
          </span>
          <span>
            #
            lifestyle
          </span>
        </div>
        <div>
          <time
            datetime="2025-08-10T10:25:58.000Z"
          >
            <p
              class="text-primary/50"
            >
              This advert was uploaded
               
              7 days
               ago.
            </p>
          </time>
        </div>
      </div>
    </div>
    <div
      class="fixed top-11.5 left-1/2 z-50 -translate-x-1/2"
    />
  </div>
</div>
`;

exports[`AdvertPage > should dispatch advert details 1`] = `
<div>
  <div
    class="wrapper only:mt-8"
  >
    <h2>
      Advert Details
    </h2>
    <div
      class="flex flex-col gap-4 md:flex-row"
    >
      <div
        class="md:flex-1"
      >
        <div
          class="mb-4 h-[370px] overflow-hidden rounded-lg bg-[#384152]"
        >
          <img
            alt="Product Image"
            class="h-full w-full object-cover"
            src="/src/assets/placeholder.webp"
          />
        </div>
        <div
          class="flex items-center justify-end"
        >
          <button
            class="btn btn-destructive"
          >
            <span
              class="-mb-0.75"
            >
              Delete this advert
            </span>
          </button>
        </div>
      </div>
      <div
        class="rounded-lg border border-neutral-500/15 p-4 md:flex-1"
      >
        <h3>
          test
        </h3>
        <div
          class="divide-b mb-2 flex gap-2 pb-2"
        >
          <span
            class="font-bold"
          >
            Price:
          </span>
          <span
            class="font-bold text-teal-600"
          >
            $
            0.00
          </span>
        </div>
        <div
          class="divide-b mb-2 flex gap-2 pb-2"
        >
          <span
            class="font-bold"
          >
            Wants to:
          </span>
          <span
            class=""
          >
            Buy
          </span>
        </div>
        <div
          class="divide-b mb-2 flex flex-wrap gap-x-2 gap-y-0.5 pb-2"
        >
          <span
            class="font-bold"
          >
            Tags:
          </span>
          <span>
            #
            work
          </span>
          <span>
            #
            lifestyle
          </span>
        </div>
        <div>
          <time
            datetime="2025-08-10T10:25:58.000Z"
          >
            <p
              class="text-primary/50"
            >
              This advert was uploaded
               
              7 days
               ago.
            </p>
          </time>
        </div>
      </div>
    </div>
    <div
      class="fixed top-11.5 left-1/2 z-50 -translate-x-1/2"
    />
  </div>
</div>
`;

exports[`AdvertPage > should dispatch alert component with error message 1`] = `
<div>
  <div
    class="wrapper only:mt-8"
  >
    <h2>
      Advert Details
    </h2>
    <div
      class="flex flex-col gap-4 md:flex-row"
    >
      <div
        class="md:flex-1"
      >
        <div
          class="mb-4 h-[370px] overflow-hidden rounded-lg bg-[#384152]"
        >
          <img
            alt="Product Image"
            class="h-full w-full object-cover"
            src="/src/assets/placeholder.webp"
          />
        </div>
        <div
          class="flex items-center justify-end"
        >
          <button
            class="btn btn-destructive"
          >
            <span
              class="-mb-0.75"
            >
              Delete this advert
            </span>
          </button>
        </div>
      </div>
      <div
        class="rounded-lg border border-neutral-500/15 p-4 md:flex-1"
      >
        <h3>
          test
        </h3>
        <div
          class="divide-b mb-2 flex gap-2 pb-2"
        >
          <span
            class="font-bold"
          >
            Price:
          </span>
          <span
            class="font-bold text-teal-600"
          >
            $
            0.00
          </span>
        </div>
        <div
          class="divide-b mb-2 flex gap-2 pb-2"
        >
          <span
            class="font-bold"
          >
            Wants to:
          </span>
          <span
            class=""
          >
            Buy
          </span>
        </div>
        <div
          class="divide-b mb-2 flex flex-wrap gap-x-2 gap-y-0.5 pb-2"
        >
          <span
            class="font-bold"
          >
            Tags:
          </span>
          <span>
            #
            work
          </span>
          <span>
            #
            lifestyle
          </span>
        </div>
        <div>
          <time
            datetime="2025-08-10T10:25:58.000Z"
          >
            <p
              class="text-primary/50"
            >
              This advert was uploaded
               
              7 days
               ago.
            </p>
          </time>
        </div>
      </div>
    </div>
    <div
      class="fixed top-11.5 left-1/2 z-50 -translate-x-1/2"
    >
      <div
        class="alert alert-error"
        role="alert"
      >
        <div
          class="flex items-center justify-between gap-1"
        >
          <div
            class="flex items-center justify-center"
          />
          <span
            class="-mb-1"
          >
            Error message
          </span>
        </div>
        <button
          aria-label="Close"
          class="flex cursor-pointer items-center justify-center rounded-lg"
          type="button"
        >
          <span
            class="sr-only"
          >
            Close
          </span>
          <span
            class="material-symbols-outlined !text-xl !font-medium"
          >
            close
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`AdvertPage > should dispatch loader when pending 1`] = `
<div>
  <div
    class="wrapper only:mt-8"
  >
    <h2>
      Advert Details
    </h2>
    <div
      class="fixed top-0 left-0 z-999 block h-full w-full bg-white"
    >
      <div
        class="absolute top-2/4 left-1/2 -translate-x-1/2 -translate-y-1/2"
        role="status"
      >
        <svg
          aria-hidden="true"
          class="inline h-18 w-18 animate-spin"
          fill="none"
          viewBox="0 0 100 101"
          xmlns="http://www.w3.org/2000/svg"
        >
          <lineargradient
            id="gradient"
            x2="1"
            y2="1"
          >
            <stop
              offset="0%"
              stop-color="#49c8ff"
            />
            <stop
              offset="100%"
              stop-color="#bd34fe"
            />
          </lineargradient>
          <path
            d="M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z"
            fill="url(#gradient)"
          />
          <path
            d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z"
            fill="currentFill"
          />
        </svg>
        <span
          class="sr-only"
        >
          Loading...
        </span>
      </div>
    </div>
    <div
      class="flex flex-col gap-4 md:flex-row"
    >
      <div
        class="md:flex-1"
      >
        <div
          class="mb-4 h-[370px] overflow-hidden rounded-lg bg-[#384152]"
        >
          <img
            alt="Product Image"
            class="h-full w-full object-cover"
            src="/src/assets/placeholder.webp"
          />
        </div>
        <div
          class="flex items-center justify-end"
        >
          <button
            class="btn btn-destructive"
          >
            <span
              class="-mb-0.75"
            >
              Delete this advert
            </span>
          </button>
        </div>
      </div>
      <div
        class="rounded-lg border border-neutral-500/15 p-4 md:flex-1"
      >
        <h3>
          test
        </h3>
        <div
          class="divide-b mb-2 flex gap-2 pb-2"
        >
          <span
            class="font-bold"
          >
            Price:
          </span>
          <span
            class="font-bold text-teal-600"
          >
            $
            0.00
          </span>
        </div>
        <div
          class="divide-b mb-2 flex gap-2 pb-2"
        >
          <span
            class="font-bold"
          >
            Wants to:
          </span>
          <span
            class=""
          >
            Buy
          </span>
        </div>
        <div
          class="divide-b mb-2 flex flex-wrap gap-x-2 gap-y-0.5 pb-2"
        >
          <span
            class="font-bold"
          >
            Tags:
          </span>
          <span>
            #
            work
          </span>
          <span>
            #
            lifestyle
          </span>
        </div>
        <div>
          <time
            datetime="2025-08-10T10:25:58.000Z"
          >
            <p
              class="text-primary/50"
            >
              This advert was uploaded
               
              7 days
               ago.
            </p>
          </time>
        </div>
      </div>
    </div>
    <div
      class="fixed top-11.5 left-1/2 z-50 -translate-x-1/2"
    />
  </div>
</div>
`;
